### PR TITLE
Hauki 18 non public resources

### PR DESCRIPTION
**Pull request by Rikuoja**
Not sure about this. Had to move permission checking to queryset level to support filtered GET queries, so perhaps better do all permission checking with the same method this way, instead of having several different methods.

Also, permissions of child resources are tricky, because we want parent resources' organization to automatically override both resource is_public as well as which users have the right to edit the children. But perhaps it kinda makes sense that resource organization is *only* relevant if it is an independent resource, with no parents.

Also added all sub organizations to any user that has a higher-level organization.

### CODE REVIEW

File: hours/permissions.py [hunk 1]
```diff
@@ -1,14 +1,77 @@
...
+    # multiple levels of parents will make the check slightly more complex
+    while parents:
+        has_request_organization |= Q(
+            **{path_to_resource + "parents__organization__in": users_organizations}
+        )
+        is_public &= Q(**{path_to_resource + "parents__isnull": True}) | Q(
+            **{path_to_resource + "parents__is_public": True}
+        )
+        parents = Resource.objects.filter(children__in=parents)
+        path_to_resource += "parents__"
```
Comment (mikkokeskinen) [inline hours/permissions.py:60]: How many queries this generates if there are like 1000 resources with parents?

**Review by mikkokeskinen — COMMENTED** [review]
This is indeed a bit tricky problem. I added a couple of comments.

File: hours/permissions.py (hunk 1 above)
Comment (mikkokeskinen) [reply hours/permissions.py:60]: I think this could be maybe achieved with a clever CTE query.

Comment (Rikuoja) [reply hours/permissions.py:60]: Surprisingly, only a few, looks like django knows how to optimize the queries done with a single Q object.

However, the single queries are a bit on the heavy side, perhaps we don't want a number of slow queries whenever the resource list is accessed? Also, here we are counting on the resource hierarchy not having many levels, each level will add two additional queries I believe.

This makes the request run for over a second for 10 resources, and three seconds for 1000 resources, so the query is slow but scaling is good. In the latter case, lots of the time is spent in json rendering tho.

These are the four slow queries, publicity and parents' publicity:

```
2020-11-27 11:00:49,206 django.db.backends DEBUG: (0.215) SELECT "hours_resource"."id", "hours_resource"."created", "hours_resource"."modified", "hours_resource"."is_removed", "hours_resource"."name", "hours_resource"."name_fi", "hours_resource"."name_sv", "hours_resource"."name_en", "hours_resource"."description", "hours_resource"."description_fi", "hours_resource"."description_sv", "hours_resource"."description_en", "hours_resource"."address", "hours_resource"."address_fi", "hours_resource"."address_sv", "hours_resource"."address_en", "hours_resource"."resource_type", "hours_resource"."organization_id", "hours_resource"."last_modified_by_id", "hours_resource"."extra_data", "hours_resource"."is_public" FROM "hours_resource" INNER JOIN "hours_resource_children" ON ("hours_resource"."id" = "hours_resource_children"."from_resource_id") WHERE (NOT "hours_resource"."is_removed" AND "hours_resource_children"."to_resource_id" IN (SELECT U0."id" FROM "hours_resource" U0 WHERE NOT U0."is_removed")); args=()
2020-11-27 11:00:49,980 django.db.backends DEBUG: (0.081) SELECT "hours_resource"."id", "hours_resource"."created", "hours_resource"."modified", "hours_resource"."is_removed", "hours_resource"."name", "hours_resource"."name_fi", "hours_resource"."name_sv", "hours_resource"."name_en", "hours_resource"."description", "hours_resource"."description_fi", "hours_resource"."description_sv", "hours_resource"."description_en", "hours_resource"."address", "hours_resource"."address_fi", "hours_resource"."address_sv", "hours_resource"."address_en", "hours_resource"."resource_type", "hours_resource"."organization_id", "hours_resource"."last_modified_by_id", "hours_resource"."extra_data", "hours_resource"."is_public" FROM "hours_resource" INNER JOIN "hours_resource_children" ON ("hours_resource"."id" = "hours_resource_children"."from_resource_id") WHERE (NOT "hours_resource"."is_removed" AND "hours_resource_children"."to_resource_id" IN (SELECT V0."id" FROM "hours_resource" V0 INNER JOIN "hours_resource_children" V1 ON (V0."id" = V1."from_resource_id") WHERE (NOT V0."is_removed" AND V1."to_resource_id" IN (SELECT U0."id" FROM "hours_resource" U0 WHERE NOT U0."is_removed")))); args=()
```

and organization and parents' organization:

```
2020-11-27 11:41:03,037 django.db.backends DEBUG: (0.143) SELECT "hours_resource"."id", "hours_resource"."created", "hours_resource"."modified", "hours_resource"."is_removed", "hours_resource"."name", "hours_resource"."name_fi", "hours_resource"."name_sv", "hours_resource"."name_en", "hours_resource"."description", "hours_resource"."description_fi", "hours_resource"."description_sv", "hours_resource"."description_en", "hours_resource"."address", "hours_resource"."address_fi", "hours_resource"."address_sv", "hours_resource"."address_en", "hours_resource"."resource_type", "hours_resource"."organization_id", "hours_resource"."last_modified_by_id", "hours_resource"."extra_data", "hours_resource"."is_public" FROM "hours_resource" LEFT OUTER JOIN "hours_resource_children" ON ("hours_resource"."id" = "hours_resource_children"."to_resource_id") LEFT OUTER JOIN "hours_resource" T3 ON ("hours_resource_children"."from_resource_id" = T3."id") WHERE (NOT "hours_resource"."is_removed" AND (("hours_resource"."is_public" AND ("hours_resource_children"."from_resource_id" IS NULL OR T3."is_public")) OR ("hours_resource_children"."from_resource_id" IS NULL AND "hours_resource"."organization_id" IN ('tprek:e58f12f4-bd80-4b4f-94ab-63ce464b6e6e', 'tprek:43fec988-487e-4eb9-8655-a0ea3ac1813f', 'tprek:272fae56-27ed-4fe4-b36d-aa6092158d26', 'tprek:36c71b04-394b-4015-bdda-c7bdf3572760', 'tprek:44916bd9-058c-43f2-a676-8f69a0b5d69b', 'tprek:628566c6-d210-4732-8f6e-0dab3761bcbe', 'tprek:6bc1d12f-88cc-46c0-a5b0-bd8286082e12', 'tprek:c087cdd4-b1aa-40e9-a122-079f44e486ff', 'tprek:6cadd9f7-7db5-456e-b5eb-4414fa6ad1de', 'tprek:02bc6a79-75d7-4725-b406-ddb210447704', 'tprek:a5dcb930-f1cc-4bb6-9f8e-c5dddda75a0b', 'tprek:18b157ac-e863-4482-8bd7-3c737e6fe6d6', 'tprek:3472e197-7dc8-43d5-bf0e-760102d01b21',...
2020-11-27 11:45:28,160 django.db.backends DEBUG: (0.072) SELECT COUNT(*) AS "__count" FROM "hours_resource" LEFT OUTER JOIN "hours_resource_children" ON ("hours_resource"."id" = "hours_resource_children"."to_resource_id") LEFT OUTER JOIN "hours_resource" T3 ON ("hours_resource_children"."from_resource_id" = T3."id") WHERE (NOT "hours_resource"."is_removed" AND (("hours_resource"."is_public" AND ("hours_resource_children"."from_resource_id" IS NULL OR T3."is_public")) OR ("hours_resource_children"."from_resource_id" IS NULL AND "hours_resource"."organization_id" IN ('tprek:e58f12f4-bd80-4b4f-94ab-63ce464b6e6e', 'tprek:43fec988-487e-4eb9-8655-a0ea3ac1813f', 'tprek:272fae56-27ed-4fe4-b36d-aa6092158d26', 'tprek:36c71b04-394b-4015-bdda-c7bdf3572760', 'tprek:44916bd9-058c-43f2-a676-8f69a0b5d69b', 'tprek:628566c6-d210-4732-8f6e-0dab3761bcbe', 'tprek:6bc1d12f-88cc-46c0-a5b0-bd8286082e12', 'tprek:c087cdd4-b1aa-40e9-a122-079f44e486ff', 'tprek:6cadd9f7-7db5-456e-b5eb-4414fa6ad1de', 'tprek:02bc6a79-75d7-4725-b406-ddb210447704', 'tprek:a5dcb930-f1cc-4bb6-9f8e-c5dddda75a0b', 'tprek:18b157ac-e863-4482-8bd7-3c737e6fe6d6', 'tprek:3472e197-7dc8-43d5-bf0e-760102d01b21', 'tprek:e3013da8-64fa-47ae-aaa7-e4ca696e5631', 'tprek:3412507f-69f2-4476-83ed-2fa57a2a8939', 'tprek:186c981b-1905-41d0-ac12-68b8c7a42b66', 'tprek:4041b1ce-6e84-477e-b51d-b49a090f539b', ...
```

File: users/models.py [hunk 2]
```diff
@@ -11,8 +13,17 @@ def __str__(self):
...
+        ):
+            return set()
+        # regular users have rights to all organizations below their level
+        orgs = []
+        for org in self.admin_organizations.all():
+            orgs.append(org.get_descendants(include_self=True))
+        for org in self.organization_memberships.all():
+            orgs.append(org.get_descendants(include_self=True))
+        # for multiple orgs, we have to combine the querysets
+        return set(reduce(lambda a, b: a | b, orgs))
```
[the remark this answers is not in the document]
Comment (Rikuoja) [reply users/models.py:29]: Good point, this combines the querysets before creating the set but better do the other way around.

File: hours/permissions.py [hunk 3]
```diff
@@ -1,14 +1,77 @@
...
+            date_periods__time_span_groups__rules__in=queryset
+        )
+        path_to_resource = "group__period__resource__"
+    elif queryset.model == TimeSpan:
+        children_to_check = Resource.objects.filter(
+            date_periods__time_span_groups__time_spans__in=queryset
+        )
+        path_to_resource = "group__period__resource__"
+    else:
+        raise Exception("Permissions not defined for model %s" % queryset.model)
```
[the remark this answers is not in the document]
Comment (Rikuoja) [reply hours/permissions.py:40]: Will fix.

**Review by mikkokeskinen — APPROVED** [review]
LGTM!

**Pull request merged**